src/dialects/abstract/index.ts

```diff
@@ -225,23 +323,132 @@ export abstract class AbstractDialect {
+  abstract readonly dataTypesDocumentationUrl: string;
+
+  readonly name: Dialect;
+  readonly DataTypes: Record<string, Class<AbstractDataType<any>>>;
+
+  /** dialect-specific implementation of shared data types */
+  #dataTypeOverrides: Map<string, Class<AbstractDataType<any>>>;
+  /** base implementations of shared data types */
+  #baseDataTypes: Map<string, Class<AbstractDataType<any>>>;
+  #dataTypeParsers = new Map<unknown, TypeParser>();
 
   get supports(): DialectSupports {
     const Dialect = this.constructor as typeof AbstractDialect;
 
     return Dialect.supports;
   }
 
+  constructor(sequelize: Sequelize, dialectDataTypes: Record<string, Class<AbstractDataType<any>>>, dialectName: Dialect) {
+    this.sequelize = sequelize;
+    this.DataTypes = dialectDataTypes;
+    this.name = dialectName;
+
+    const baseDataTypes = new Map<string, Class<AbstractDataType<any>>>();
+    for (const dataType of Object.values(BaseDataTypes) as Array<Class<AbstractDataType<any>>>) {
+      const dataTypeId: string = (dataType as unknown as typeof AbstractDataType).getDataTypeId();
+
+      // intermediary data type
+      if (!dataTypeId) {
+        continue;
+      }
+
+      if (baseDataTypes.has(dataTypeId)) {
+        throw new Error(`Internal Error: Sequelize declares more than one base implementation for DataType ID ${dataTypeId}.`);
+      }
+
+      baseDataTypes.set(dataTypeId, dataType);
+    }
+
+    const dataTypeOverrides = new Map<string, Class<AbstractDataType<any>>>();
+    for (const dataType of Object.values(this.DataTypes)) {
+      const replacedDataTypeId: string = (dataType as unknown as typeof AbstractDataType).getDataTypeId();
+
+      if (dataTypeOverrides.has(replacedDataTypeId)) {
+        throw new Error(`Dialect ${this.name} declares more than one implementation for DataType ID ${replacedDataTypeId}.`);
+      }
+
+      dataTypeOverrides.set(replacedDataTypeId, dataType);
+    }
+
+    this.#dataTypeOverrides = dataTypeOverrides;
+    this.#baseDataTypes = baseDataTypes;
+  }
+
+  /**
+   * Returns the dialect-specific implementation of a shared data type, or null if no such implementation exists
+   * (in which case you need to use the base implementation).
+   *
+   * @param dataType The shared data type.
+   */
+  getDataTypeForDialect(dataType: Class<AbstractDataType<any>>): Class<AbstractDataType<any>> | null {
+    const typeId = (dataType as unknown as typeof AbstractDataType).getDataTypeId();
+    const baseType = this.#baseDataTypes.get(typeId);
+
+    // this is not one of our types. May be a custom type by a user. We don't replace it.
+    if (baseType != null && baseType !== dataType) {
+      return null;
+    }
+
+    return this.#dataTypeOverrides.get(typeId) ?? null;
+  }
+
+  #printedWarnings = new Set<string>();
+  warnDataTypeIssue(text: string): void {
+    // TODO: log this to sequelize's log option instead (requires a logger with multiple log levels first)
```

Comment:
I think [Winston](https://github.com/winstonjs/winston#logging-levels) will good for this.

Reply:
Or if we want the newest generation for the logger option with healthy maintenance status, we can use [pino](https://getpino.io/#/docs/help?id=log-levels-as-labels-instead-of-number).

Reply:
We'll discuss that when we start working on it :)